dips/dip-213.md

```diff
@@ -0,0 +1,94 @@
+
+```
+Current paradigm: Order → Execute → Commit → Order → Execute → Commit
+
+Proposed paradigm: Order → Order → Order
+                     ↓        ↓
+                     Execute → Execute
+                        ↓        ↓
+                       Commit → Commit
+```
+
+If the time of ordering, execution and commit for a block is O, E, C correspondingly, the throughput should change from 1/(O + E + C) to 1/min(O, E, C) which not only unlocks throughput benefits immediately,
+but makes bottlenecks more visible and could guide future optimizations.
+
+# Description
+
+In [DiemBFTv4](https://developers.diem.com/papers/diem-consensus-state-machine-replication-in-the-diem-blockchain/2021-08-17.pdf), a block (proposal) goes through different steps before it's finalized. The steps are 
+```
+Consensus stage
+    1. Proposed
+    2. Executed
+    3. Voted (including execution result)
+    4. QuorumCertified
+    5. 2-chain Certified (Finalized)
+```
+
+The proposed change is to decouple execution from consensus and pipeline the process with different stages. A block would go through
+4 stages each with its own steps:
+```
+1. Consensus stage 
+    * Proposed
+    * Voted (no execution result)
+    * QuorumCertified
+    * 2-chain Certified (Ordered)
+2. Execution stage
+3. Voting stage (for execution result)
+4. Commit stage (Finalized)
+```
+The stages run in parallel, utilizing available compute resources to increase overall throughput. 
+That is, the system could commit B1, sign B2, execute B3, and order B4, all at the same time. Note, while throughput increases due to parallel execution, latency of an individual transaction (block) finalization does not decrease, as the stages are inherently sequential. In fact, individual transaction latency may slightly increase since commit occurs less "eagerly".
+
+# Required Changes
+
+## Consensus
+`StateComputer` is the trait that consensus uses for `compute`, `commit` or `sync_to` blocks.
+An `ordering_state_computer` is implemented to bypass execution and send blocks to next stage when ordered.
+
+## BufferManager
+A `BufferManager` is implemented to manage different stages of ordered blocks, it's a queue of blocks with stage markers.
+- `Ordered` block is sent to execution and advanced to `Executed` after receiving execution result.
+- `Executed` block is sent to safety rules and advanced to `Signed` after receiving a signature.
+- Signature on `Signed` block is broadcasted to all validators. At all stages except `Aggregated`, signatures from other validators on the block are collected.
+- A block in either `Executed` or `Signed` stage is advanced to `Aggregated` after receiving a quorum of signatures.
+- `Aggregated` block is popped from the queue and sent to persistent storage.
+ 
+## Sync
+When the node is far from current proposed block, it may decide to fast-forward via state sync protocol. It's triggered if the gap between
```

Comment:
```suggestion
When a node's local blockchain tail is far from current proposed block, it may decide to fast-forward via state sync protocol. It's triggered if the gap between
```

Reply:
what do you mean by "the node is far"? do you mean it local blockchain tail?

and how does the node learn about the "gap" without being in state-sync already? this seems circular, first you need to state-sync in order to discover the gap and then initiate a state-sync ..

Reply:
the node learns about the gap by checking round numbers from SyncInfo (all highest certs) (which is carried in proposal/vote/timeout etc) and decides whether to initiate state-sync

Reply:
got it. please incorporate the explanation here.